// kafkajs/src/protocol/requests/index.js
const apiKeys = require('kafkajs/src/protocol/requests/apiKeys')
const { KafkaJSServerDoesNotSupportApiKey } = require('kafkajs/src/errors')

const requests = {
  Produce: require('kafkajs/src/protocol/requests/produce'),
  Fetch: require('kafkajs/src/protocol/requests/fetch'),
  ListOffsets: require('kafkajs/src/protocol/requests/listOffsets'),
  Metadata: require('kafkajs/src/protocol/requests/metadata'),
  LeaderAndIsr: {},
  StopReplica: {},
  UpdateMetadata: {},
  ControlledShutdown: {},
  OffsetCommit: require('kafkajs/src/protocol/requests/offsetCommit'),
  OffsetFetch: require('kafkajs/src/protocol/requests/offsetFetch'),
  GroupCoordinator: require('kafkajs/src/protocol/requests/findCoordinator'),
  JoinGroup: require('kafkajs/src/protocol/requests/joinGroup'),
  Heartbeat: require('kafkajs/src/protocol/requests/heartbeat'),
  LeaveGroup: require('kafkajs/src/protocol/requests/leaveGroup'),
  SyncGroup: require('kafkajs/src/protocol/requests/syncGroup'),
  DescribeGroups: require('kafkajs/src/protocol/requests/describeGroups'),
  ListGroups: {},
  SaslHandshake: require('kafkajs/src/protocol/requests/saslHandshake'),
  ApiVersions: require('kafkajs/src/protocol/requests/apiVersions'),
  CreateTopics: require('kafkajs/src/protocol/requests/createTopics'),
  DeleteTopics: require('kafkajs/src/protocol/requests/deleteTopics'),
  DeleteRecords: {},
  InitProducerId: require('kafkajs/src/protocol/requests/initProducerId'),
  OffsetForLeaderEpoch: {},
  AddPartitionsToTxn: require('kafkajs/src/protocol/requests/addPartitionsToTxn'),
  AddOffsetsToTxn: require('kafkajs/src/protocol/requests/addOffsetsToTxn'),
  EndTxn: require('kafkajs/src/protocol/requests/endTxn'),
  WriteTxnMarkers: {},
  TxnOffsetCommit: require('kafkajs/src/protocol/requests/txnOffsetCommit'),
  DescribeAcls: {},
  CreateAcls: {},
  DeleteAcls: {},
  DescribeConfigs: require('kafkajs/src/protocol/requests/describeConfigs'),
  AlterConfigs: require('kafkajs/src/protocol/requests/alterConfigs'),
  AlterReplicaLogDirs: {},
  DescribeLogDirs: {},
  SaslAuthenticate: require('kafkajs/src/protocol/requests/saslAuthenticate'),
  CreatePartitions: {},
  CreateDelegationToken: {},
  RenewDelegationToken: {},
  ExpireDelegationToken: {},
  DescribeDelegationToken: {},
  DeleteGroups: {},
}

const BEGIN_EXPERIMENTAL_V011_REQUEST_VERSION = {
  [apiKeys.Produce]: 3,
  [apiKeys.Fetch]: 4,
}

const names = Object.keys(apiKeys)
const keys = Object.values(apiKeys)
const findApiName = apiKey => names[keys.indexOf(apiKey)]

const lookup = (versions, allowExperimentalV011) => (apiKey, definition) => {
  const version = versions[apiKey]
  const availableVersions = definition.versions.map(Number)
  const allowedVersions = allowExperimentalV011
    ? availableVersions
    : availableVersions.filter(
        version =>
          !BEGIN_EXPERIMENTAL_V011_REQUEST_VERSION[apiKey] ||
          version < BEGIN_EXPERIMENTAL_V011_REQUEST_VERSION[apiKey]
      )
  const bestImplementedVersion = Math.max.apply(this, allowedVersions)

  if (!version || version.maxVersion == null) {
    throw new KafkaJSServerDoesNotSupportApiKey(
      `The Kafka server does not support the requested API version`,
      { apiKey, apiName: findApiName(apiKey) }
    )
  }

  const bestSupportedVersion = Math.min(bestImplementedVersion, version.maxVersion)
  return definition.protocol({ version: bestSupportedVersion })
}

module.exports = {
  requests,
  lookup,
}
